### DISCUSSION TITLE
[@types/jest] global function fail is not defined

### DISCUSSION THREAD

**Discussion by srmagura** [DefinitelyTyped/DefinitelyTyped#55803] [Issues with a @types package]
`@types/jest` [declares a global function called `fail`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L1352), but it seems `fail` no longer exists in Jest. I'm not sure if this is a problem in `@types/jest` or `jest` itself.

Minimal repro: https://github.com/srmagura/jest-fail-repro

```ts
it("fails", () => {
  fail();
});
```
--> `ReferenceError: fail is not defined`

[a comment by an automated account is not shown]

**Comment by domdomegg** [comment]
Thanks for raising this. I've dug into it a bit and found this used to work in Jest at versions <26 with the defaults as they were using the `jest-jasmine2` test runner. Now the default is to use `jest-circus`, which doesn't provide this fail method.

Instead, in Jest you should simply throw an error as this will be caught by the test runner:
- `fail()` :arrow_right: `throw new Error()`
- `fail('Message with reason')` :arrow_right: `throw new Error('Message with reason')`

You can do this across your codebase with this regex find and replace:

![2021-09-16_13-26](https://user-images.githubusercontent.com/4953590/133611876-75e52073-f904-4a93-89d1-9d23de557ae0.png)
(`fail\((.*?)\)` :arrow_right: `throw new Error($1)`)

<details>
<summary>Using the old test runner (not recommended)</summary>

<br/>To use the old test runner, you can add the configuration`"testRunner": "jest-jasmine2"` in your package.json like:

```jsonc
{
  "name": "my-package",
  "jest": {
    "testRunner": "jest-jasmine2"
  },
  ...
}
```

</details>

**Comment by domdomegg** [comment]
I guess the other question to answer here is what we do about the types. Effectively, we have different types depending on the configuration of the test runner. If we remove `fail` from the types then it'll break for anyone using the old test runner. If we keep it in, it'll confuse users like this with runtime errors.

I'm not sure what the right way to do this is. Potentially we have a new package called `@types/jest/jasmine2` or similar that is a drop in replacement for `@types/jest` but includes the jasmine runner types as well?

**Comment by srmagura** [comment]
Great feedback. A `@types/jest/jasmine2` package sounds like a good general solution.

Adding a `fail` function to `jest-circus` could resolve this specific issue. I like the brevity and readability `fail` provides over `throw new Error()`.

But I defer to the maintainers on this one.

**Comment by srmagura** [comment]
Ah, I found a Jest issue, https://github.com/facebook/jest/issues/11698, for `fail` not being defined by `jest-circus`. It seems like this was an unintentional breaking change.
